proposals/separate-get-set-field.md

```diff
@@ -0,0 +1,187 @@
+```
+
+The above when combined with the dot-syntax would allow for you to write
+`foo & l.name .~ 5` instead of `foo & l @"name" .~ 5` or `foo & l #name .~ 5`.
+
+More generally it allows you to define potentially infinite and fairly magical
+"records" which you can use clean and intuitive dot-syntax to work with.
+
+Obviously the above has absolutely no chance of supporting a reasonable
+equivalent `SetField`.
+
+## Proposed Change Specification
+
+`GHC.Records.HasField` would be adjusted from it's current planned
+specification:
+
+```haskell
+class HasField x r a | x r -> a where
+    hasField :: r -> (a -> r, a)
+```
+
+to become
+
+```haskell
+class GetField (x :: k) r a | x r -> a where
+    getField :: r -> a
+
+class GetField x r a => SetField (x :: k) r a | k r -> a where
+    setField :: a -> r -> r
+    setField = updateField @x . const
+
+    updateField :: (a -> a) -> r -> r
+    updateField f r = let (g, a) = hasField @x r in g (f a)
+
+    hasField :: r -> (a -> r, a)
+    hasField r = (\a -> setField @x a r, getField @x r)
+
+    {-# MINIMAL setField | updateField | hasField #-}
+```
+
+
+## Effect and Interactions
+
+This change would allow for the definition of various `GetField` instances
+that were mentioned above as desirable. This would then enable those instances
+to be utilized by dot-notation.
+
+This would not be backwards compatible with the existing implementation,
+but neither is the planned implementation. Unlike the planned implementation
+it would be a very mechanical rename to fix, as opposed to a potentially
+impossible definition of `hasField`.
+
+
+## Costs and Drawbacks
+
+The drawbacks seem minimal compared to the planned implementation. You can just
+treat `SetField` as the proposed `HasField` and everything will work the same.
+
+There is slightly more complexity due to having two separate classes rather
+than just one.
+
+
+## Alternatives
+
+Separate out further into "get-only", "set-only", "get-0-or-more",
+"get-and-set" etc. instead of only "get-only" and "get-and-set".
+
+The above makes a lot of sense for libraries like `lens` that deal with sums and
+clever multi-target lenses like `traverse`.
+
+However I do not believe it is particularly valuable for these `GHC.Records`
+classes, since they are focused on "fields" and "records" and automatically
+deriving them, and not on the general concept of lenses.
+
+Whenever you have a record that contains a `Maybe b` or a `[b]`, you will most
+likely want (and have derived for you) the single-target field from `a` to
+`Maybe b`/`[b]`, and not the multi/partial-target field from `a` to `b`. The
+latter you can always define yourself as a top level lens or as an instance of
+some other class.
+
+
+## Unresolved Questions
+
+Physical fields are currently resolved automatically by GHC whenever the field
+is in scope, so to allow for GHC to autoderive read-only fields we will want
+some syntax to export only the GetField instance and not the SetField instance.
```

Comment:
I think the proposal should either propose something here, or be explicit that in its current state it will support read-only or read-write virtual fields but always generate read-write physical fields. I suppose if you want a read-only field you can always do this:
```hs
newtype AsciiString = AsciiString Text

contents :: AsciiString -> Text
contents (AsciiString t) = t

instance GetField "contents" AsciiString Text where
  getField = contents
```
Is this good enough? The existing property of `HasField`, that the constraints get solved iff the field selector is in scope, has the merit of simplicity.

Reply:
Yeah requiring virtual fields for read-only fields seems fine for the time being. Similar to your example I can also do:

```
data Foo = Foo
    { _bar :: Bar
    , _baz :: Baz
    }

instance GetField "bar" Foo Bar where
    getField = ._bar

instance GetField "baz" Foo Baz where
    getField = ._baz
```

It would be nice if I could keep the Generic instance and the dot-notation syntax in sync, but it's not overly pressing.

I'll adjust that and then maybe make another proposal sometime later for that.